lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,316 @@
+          @fields.merge(mappings[:checksum] => generate_checksum)
+        end
+
+        def generate_checksum
+          
+          checksum_payload_items = Hash.new
+          CHECKSUM_FIELDS.each do |field|
+              checksum_payload_items[field] = @fields[field]
+          end          
+          Paytm.checksum(checksum_payload_items, @options[:credential2])
+        end
+
+        def sanitize_fields
+          %w(email phone).each do |field|
+            @fields[field].gsub!(/[^a-zA-Z0-9\-_@\/\s.]/, '') if @fields[field]
+          end
+        end
+
+      end
+
+
+
+      class Notification < OffsitePayments::Notification
+        PAYTM_RESPONSE_PARAMS = %w(SUBS_ID MID BANKTXNID TXNAMOUNT CURRENCY STATUS RESPCODE RESPMSG TXNDATE GATEWAYNAME BANKNAME PAYMENTMODE PROMO_CAMP_ID PROMO_STATUS PROMO_RESPCODE ORDERID TXNID REFUNDAMOUNT REFID).freeze
+        def initialize(post, options = {})
+          super(post, options)
+          @merchant_id = options[:credential1]
+          @secret_key = options[:credential2]
+        end
+
+        def complete?
+          status == 'Completed'
+        end
+
+        def status
+          if transaction_status.casecmp("TXN_SUCCESS").zero? 
+            return 'Completed'
+          elsif transaction_status.casecmp("pending").zero? 
+            return 'Pending'
+          else
+            return 'Failed'
+          end
+        end
+
+        def invoice_ok?(order_id)
+          order_id.to_s == invoice.to_s
+        end
+
+        # Order amount should be equal to gross - discount
+        def amount_ok?(order_amount, _order_discount = BigDecimal.new('0.0'))
+          BigDecimal.new(original_gross) == order_amount
+        end
+
+        # Status of transaction return from the Paytm. List of possible values:
+        # <tt>TXN_SUCCESS</tt>::
+        # <tt>pending</tt>::
+        # <tt>TXN_FAILURE</tt>::
+        def transaction_status
+          params['STATUS']
+        end
+
+        # ID of this transaction (Paytm transaction id)
+        def transaction_id
+          params['TXNID']
+        end
+
+        # Mode of Payment
+        #
+        # 'CC' for credit-card
+        # 'NB' for net-banking
+
+        def type
+          params['PAYMENTMODE']
+        end
+
+        # What currency have we been dealing with
+        def currency
+          'INR'
+        end
+
+        def item_id
+          params['ORDERID']
+        end
+
+        # This is the invoice which you passed to Paytm
+        def invoice
+          params['ORDERID']
+        end
+
+        # Merchant Id provided by the Paytm
+        def account
+          params['MID']
+        end
+
+        # original amount send by merchant
+        def original_gross
+          params['TXNAMOUNT']
+        end
+
+        def gross
+          parse_and_round_gross_amount(params['TXNAMOUNT'])
+        end
+
+        def checksum #double check check get via get call
+          params['CHECKSUMHASH'].tr(' ', '+')
```

Comment:
It's failing again because `params['CHECKSUMHASH']` is bring returning as `nil` here. 😕